docs/proposals/loadbalancing.md

```diff
@@ -0,0 +1,311 @@
+
+```yaml
+apiVersion: v1
+kind: IngressPath
+metadata:
+  name: l4ingress
+  type: network
+spec:
+  host: www.example.com
+  tlsMode: Termination
+  secret: foosecret
+status:
+  loadBalancer:
+    ingress:
+    - ip: 104.1..
+```
+
+type: network with a pathmap is a validation error.
+This adds a slight burden on users, since what was previously a one line change to expose a service now requires a new resource.
+
+
+### Why not use a single loadbalancer for all Ingress paths?
+
+Use cases for not having all ingress paths join an existing load balancer:
+* User wants only foo.com and bar.com backed by a single ip
+* Internal services get bare metal loadbalancing, user facing services are backed by a cloud loadbalancer
+* User already has a specific ip in DNS, and wants to expose a loadbalanced website on it
+* A user wants to isolate their host from decryption of other hosts saturating cpu on a single bare metal loadbalancer (this isn't really a problem for cloud) by explicitly asking for a new one.)
+
+### How does the ingress path choose a loadbalancer to join?
+
+For 1.1, there will only be one loadbalancer controller that claims all ingress paths. Whether it creates a new loadbalancer or reuses an existing one is upto the controller. For example, the GCE controller will create 1 loadbalancer per ingress path, if it has `type: network`, this will be an l4, if it has `type: application` it will be an l7 with rules to allow proxying to different backends based on the given pathmap.
+
+However we need to provide a way for users to specify which class of loadbalancers to join, and which particular loadbalancer (in the case of multiple loadbalancer per class). Note that we can't just say each class of loadbalancers has a single loadbalancer in it, and so picking a class boils down to picking a specific loadbalancer, because of cloud provider limitations. Nor can we say picking a specific loadbalancer picks the class, because a loadbalancer might have limited capacity (i.e don't force a user to pick 172.12, which might be maxed out, when they just want class=gce).
+
+The proposed long term solution for this problem is to reference a claim on an ip through the Ingress Path.
+
+### How does TLS work?
+
+Each IngressPath specifies a single secret and a tls mode. Its upto the loadbalancer controller to handle the tls mode. For example, the gce controller will only support TLS termination for 1.1 because each IngressPath gets a new loadbalancer, and GCE doesn't allow multiple certs per IP. With the current design a user can create the wrong type of secret for the wrong mode. They wouldn't know till they try creating the Ingress path. Validating a simple key:value secret is also hard.
+
+Some alternatives to the proposed tls mode handling:
```

Comment:
Even if the user provides a secret the admin may choose not to expose it, or have a policy that controls that.  I think we have to allow for the possibility that everything in spec is a request, and either represent the outcome in status or provide an alternate way to communicate that.  Ingress as a request with concrete status allows variation among installations but consistency among clients.  A particular implementation can simply not report data that is not relevant/possible